### PR TITLE
compare min and max in equota

### PR DESCRIPTION
**Pull request by tornado404**

### CODE REVIEW

File: pkg/apiserver/controller/queue/queue.go [hunk 1]
```diff
@@ -348,17 +349,25 @@ func UpdateQueue(ctx *logger.RequestContext, request *UpdateQueueRequest) (Updat
...
 		if err = schema.ValidateResourceInfo(request.MinResources, scalarResourceLaws); err != nil {
 			ctx.Logging().Errorf("update queue failed. error: %s", err.Error())
 			ctx.ErrorCode = common.InvalidComputeResource
 			return UpdateQueueResponse{}, err
 		}
 		updateClusterRequired = true
+		resourceUpdated = true
 		queueInfo.MinResources = request.MinResources
 	}
+	if resourceUpdated && !request.MinResources.LessEqual(request.MaxResources) {
```
Comment (D0m021ng) [inline pkg/apiserver/controller/queue/queue.go]: Only  elasticQuota need `minResources`

Comment (tornado404) [reply pkg/apiserver/controller/queue/queue.go]: done

File: pkg/apiserver/controller/queue/queue.go [hunk 2]
```diff
@@ -419,6 +426,32 @@ func UpdateQueue(ctx *logger.RequestContext, request *UpdateQueueRequest) (Updat
...
+	if resource.CPU != "" {
+		queueResource.CPU = resource.CPU
+	}
+	if resource.Mem != "" {
+		queueResource.Mem = resource.Mem
+	}
+	if resource.ScalarResources != nil {
+		queueResource.ScalarResources = make(schema.ScalarResourcesType)
+	}
+	if len(resource.ScalarResources) != 0 || resource.ScalarResources == nil {
```
Comment (D0m021ng) [inline pkg/apiserver/controller/queue/queue.go]: When `resource.ScalarResources == nil` ,  `range rsource.ScalarResources` might be panic.

File: pkg/apiserver/controller/queue/queue.go [hunk 3]
```diff
@@ -338,26 +338,33 @@ func UpdateQueue(ctx *logger.RequestContext, request *UpdateQueueRequest) (Updat
 
 	// validate fields if not nil, validate namespace at first
 	updateClusterRequired := false
+	resourceUpdated := true
```
Comment (D0m021ng) [inline pkg/apiserver/controller/queue/queue.go]: When does the variable `resourceUpdated` become `false`?

File: pkg/apiserver/controller/queue/queue.go [hunk 4]
```diff
@@ -338,26 +338,33 @@ func UpdateQueue(ctx *logger.RequestContext, request *UpdateQueueRequest) (Updat
...
 
 	// validate MaxResource or MinResource
-	scalarResourceLaws := config.GlobalServerConfig.Job.ScalarResourceArray
-	if request.MaxResources.CPU != "" && request.MaxResources.Mem != "" {
-		if err = schema.ValidateResourceInfo(request.MaxResources, scalarResourceLaws); err != nil {
-			ctx.Logging().Errorf("update queue failed. error: %s", err.Error())
+	if request.MinResources.CPU != "" || request.MinResources.Mem != "" {
+		if err := validateQueueResource(request.MaxResources, &queueInfo.MaxResources); err != nil {
```
Comment (D0m021ng) [inline pkg/apiserver/controller/queue/queue.go]: Line 344 checks `request.MinResources`, and Line 345 checks `request.MaxResources`

File: pkg/apiserver/controller/queue/queue.go [hunk 5]
```diff
@@ -338,26 +338,33 @@ func UpdateQueue(ctx *logger.RequestContext, request *UpdateQueueRequest) (Updat
...
+			ctx.Logging().Errorf("update queue minResources failed. error: %s", err.Error())
 			ctx.ErrorCode = common.InvalidComputeResource
 			return UpdateQueueResponse{}, err
 		}
 		updateClusterRequired = true
-		queueInfo.MinResources = request.MinResources
+		resourceUpdated = true
+	}
+	if queueInfo.QuotaType == schema.TypeElasticQuota && resourceUpdated &&
+		!request.MinResources.LessEqual(request.MaxResources) {
```
Comment (D0m021ng) [inline pkg/apiserver/controller/queue/queue.go]: If we only update `MinResources` and `MaxResources` is empty, this check might be wrong.

File: pkg/apiserver/controller/queue/queue.go (hunk 2 above)
Comment (tornado404) [reply pkg/apiserver/controller/queue/queue.go]: range nil resource.ScalarResources won't effect

File: pkg/apiserver/controller/queue/queue.go (hunk 3 above)
Comment (tornado404) [reply pkg/apiserver/controller/queue/queue.go]: resourceUpdated should be false

File: pkg/apiserver/controller/queue/queue.go (hunk 4 above)
Comment (tornado404) [reply pkg/apiserver/controller/queue/queue.go]: repaired

File: pkg/apiserver/controller/queue/queue.go (hunk 5 above)
Comment (tornado404) [reply pkg/apiserver/controller/queue/queue.go]: agree it

File: pkg/apiserver/controller/queue/queue.go [hunk 6]
```diff
@@ -337,43 +337,51 @@ func UpdateQueue(ctx *logger.RequestContext, request *UpdateQueueRequest) (Updat
...
-	} else {
+	} else if request.Location != nil {
+		updateClusterRequired = true
 		log.Debugf("queue %s Location is set nil", request.Name)
 	}
 
 	// validate scheduling policy
 	if len(request.SchedulingPolicy) != 0 {
 		log.Warningf("todo queue.SchedulingPolicy havn't been validated yet")
+		updateClusterRequired = true
```
Comment (D0m021ng) [inline pkg/apiserver/controller/queue/queue.go]: Why do you need update `SchedulingPolicy` on cluster?

File: pkg/apiserver/controller/queue/queue.go [hunk 7]
```diff
@@ -337,43 +337,51 @@ func UpdateQueue(ctx *logger.RequestContext, request *UpdateQueueRequest) (Updat
...
-		queueInfo.MinResources = request.MinResources
 	}
+
 	// validate Location
 	if queueInfo.Location == nil {
 		queueInfo.Location = make(map[string]string)
 	}
-	if len(request.Location) != 0 || request.Location == nil {
+	if len(request.Location) != 0 {
+		updateClusterRequired = true
```
Comment (D0m021ng) [inline pkg/apiserver/controller/queue/queue.go]: Why do you need update `Location` on cluster?

**Review by D0m021ng — APPROVED** [review]
LGTM

**Pull request merged**